trajectory_executor/com_serial.py

```diff
@@ -0,0 +1,222 @@
+          data_packet['r_knee'] += decoded_data
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_L_SHOULDER']
+        elif read_state == serial_read_states['READ_L_SHOULDER']:
+          data_packet['l_shoulder'] += decoded_data
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_L_ELBOW']
+        elif read_state == serial_read_states['READ_L_ELBOW']:
+          data_packet['l_elbow'] += decoded_data
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_R_SHOULDER']
+        elif read_state == serial_read_states['READ_R_SHOULDER']:
+          data_packet['r_shoulder'] += decoded_data
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_R_ELBOW']
+        elif read_state == serial_read_states['READ_R_ELBOW']:
+          data_packet['r_elbow'] += decoded_data
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_AT_GOAL']
+        elif read_state == serial_read_states['READ_AT_GOAL']:
+          data_packet['at_goal'] = decoded_data
+          calculated_checksum += decoded_data
+          read_state = serial_read_states['READ_CHECKSUM']
+        elif read_state == serial_read_states['READ_CHECKSUM']:
+          data_packet['checksum'] = decoded_data
+          if data_packet['checksum'] == 255 - (calculated_checksum % 256):
+            # good packet, relay to tracking node
+            response.roll = data_packet['roll']
+            response.pitch = data_packet['pitch']
+            response.yaw = data_packet['yaw']
+            response.l_hip = data_packet['l_hip']
+            response.l_knee = data_packet['l_hip']
+            response.r_hip = data_packet['r_hip']
+            response.r_knee = data_packet['r_hip']
+            response.l_shoulder = data_packet['l_shoulder']
+            response.l_elbow = data_packet['l_elbow']
+            response.r_shoulder = data_packet['r_shoulder']
+            response.r_elbow = data_packet['r_elbow']
+            response.at_goal = data_packet['at_goal']
+            break
+          else:
+            # bad packet, retry request
+            return self.sensor_data_request_callback(request, response)
```

Comment:
Would this default behavior cause any timing issues? Particularly, I'm just concerned if a packet fails and in retrying, we end up backing the buffer. Again, not an expert by any means so if that's not how things work, lmk!

Reply:
Yea I understand. I was skeptical even when adding this in. It's actually not fully implemented. My thought process for this was to have a number of attempts that it tracked and bails out if it exceeds say 5 tries. Might be better to just bail out immediately considering we'll probably be running the service call pretty frequently.

Reply:
I think we should log that we received a bad packet. Also let's have a limit on retry like you suggested Andrew. We should define retry behavior for both ros-sim and ros-realrobot. These should be identical so it would be best if we document that. **Maybe just add the logging here itself?**

Reply:
Yea we can just log it from here. Regarding the retries, after thinking about it more I think it would be better to just have it bail out rather than block the process to retry. We'll be calling the service frequently enough that some errors shouldn't be an issue. Even then, I doubt we'll have a significant number of errors.